syntax = "proto3";
package cortex.ml;

service MlService {
  rpc EmbedBatch (EmbedRequest) returns (EmbedResponse);
  rpc Rerank (RerankRequest) returns (RerankResponse);
  rpc Generate (GenerateRequest) returns (stream GenerateChunk);
}

message EmbedRequest {
  repeated string texts = 1;
  string model = 2;
}

message EmbedResponse {
  repeated Embedding embeddings = 1;
  int32 dimensions = 2;
}

message Embedding {
  repeated float values = 1;
}

message RerankRequest {
  string query = 1;
  repeated RerankDocument documents = 2;
  int32 top_k = 3;
  string model = 4;
}

message RerankDocument {
  string id = 1;
  string text = 2;
  float score = 3;
  map<string, string> metadata = 4;
}

message RerankResponse {
  repeated RerankDocument documents = 1;
}

message GenerateRequest {
  string prompt = 1;
  string system_prompt = 2;
  string provider = 3;
  string model = 4;
  float temperature = 5;
  int32 max_tokens = 6;
}

message GenerateChunk {
  string text = 1;
  bool is_final = 2;
  repeated Citation citations = 3;
}

message Citation {
  string chunk_id = 1;
  string document_title = 2;
  string source_url = 3;
  string snippet = 4;
  float relevance_score = 5;
}
